felix/nftables/table.go

```diff
@@ -0,0 +1,1173 @@
+
+		// Mangle hooks.
+		"mangle-PREROUTING":  {Name: "mangle-PREROUTING", Hook: &preroutingHook, Type: &filterType, Priority: &manglePriority},
+		"mangle-INPUT":       {Name: "mangle-INPUT", Hook: &inputHook, Type: &filterType, Priority: &manglePriority},
+		"mangle-FORWARD":     {Name: "mangle-FORWARD", Hook: &forwardHook, Type: &filterType, Priority: &manglePriority},
+		"mangle-OUTPUT":      {Name: "mangle-OUTPUT", Hook: &outputHook, Type: &routeType, Priority: &manglePriority},
+		"mangle-POSTROUTING": {Name: "mangle-POSTROUTING", Hook: &postroutingHook, Type: &filterType, Priority: &manglePriority},
+
+		// Raw hooks.
+		"raw-PREROUTING": {Name: "raw-PREROUTING", Hook: &preroutingHook, Type: &filterType, Priority: &rawPriority},
+		"raw-OUTPUT":     {Name: "raw-OUTPUT", Hook: &outputHook, Type: &filterType, Priority: &rawPriority},
+	}
+
+	// Prometheus metrics.
+	countNumRestoreCalls = prometheus.NewCounter(prometheus.CounterOpts{
+		Name: "felix_nft_calls",
+		Help: "Number of nft calls.",
+	})
+	countNumRestoreErrors = prometheus.NewCounter(prometheus.CounterOpts{
+		Name: "felix_nft_errors",
+		Help: "Number of nft errors.",
+	})
+	countNumSaveCalls = prometheus.NewCounter(prometheus.CounterOpts{
+		Name: "felix_nft_list_calls",
+		Help: "Number of nft list calls.",
+	})
+	countNumSaveErrors = prometheus.NewCounter(prometheus.CounterOpts{
+		Name: "felix_nft_list_errors",
+		Help: "Number of nft list errors.",
+	})
+	gaugeNumChains = prometheus.NewGaugeVec(prometheus.GaugeOpts{
+		Name: "felix_nft_chains",
+		Help: "Number of active nft chains.",
+	}, []string{"ip_version", "table"})
+	gaugeNumRules = prometheus.NewGaugeVec(prometheus.GaugeOpts{
+		Name: "felix_nft_rules",
+		Help: "Number of active nftables rules.",
+	}, []string{"ip_version", "table"})
+	countNumLinesExecuted = prometheus.NewCounterVec(prometheus.CounterOpts{
+		Name: "felix_nftables_lines_executed",
+		Help: "Number of nftables rule updates executed.",
+	}, []string{"ip_version", "table"})
+)
+
+func init() {
+	prometheus.MustRegister(countNumRestoreCalls)
+	prometheus.MustRegister(countNumRestoreErrors)
+	prometheus.MustRegister(countNumSaveCalls)
+	prometheus.MustRegister(countNumSaveErrors)
+	prometheus.MustRegister(gaugeNumChains)
+	prometheus.MustRegister(gaugeNumRules)
+	prometheus.MustRegister(countNumLinesExecuted)
+}
+
+// nftablesTable is an implementation of the generictables.Table interface that programs nftables. It represents a
+// single nftables table.
+type nftablesTable struct {
+	common.IPSetsDataplane
+
+	name      string
+	ipVersion uint8
+	nft       knftables.Interface
+
+	// For rendering rules and chains.
+	render generictables.NFTRenderer
+
+	// featureDetector detects the features of the dataplane.
+	featureDetector environment.FeatureDetectorIface
+
+	// TODO: There are no kernel chains in nftables. We should simplify this for nftables.
```

Comment:
> With no kernel chains to share, is there a way for us to nicely interleave with user-provided rules; for example, our inserted and appended rules could map to two different priorities?

"well-defined priorities" is what we were planning for kube-proxy. (btw, @caseydavenport did you end up needing to depend on any assumptions about how kube-proxy behaves anywhere? I'm on the hook to document that for 1.33...)

Reply:
I haven't thought much about it - I suspect different priorities would do the trick here.

For "deny" rules, users can just put them after all of our rules since later tables still get a chance to deny traffic even if our table accepted it. But yeah, for other types of interaction might want separate priorities. Will raise a task to track this.

Reply:
@danwinship regarding your comment (which magically appeared after my last comment) - I haven't needed to do anything special to interact with kube-proxy so far, which is great. Simply using the default / recommended priorities for each hook has worked well.